### PR TITLE
Tile Index creation: add multithreaded support, where query support, new options

### PR DESCRIPTION
**Pull request by ibell13**
Added new options for creating a tile index with the _pdal tindex create_ command:
- _threads_/_requests_ option: specify the number of CPU threads used to create point cloud boundaries in a new tile index; default 1
- _where_ option: PDAL where expression to run on the points being indexed
- _simplify_, _threshold_, _resolution_ and _sample_size_ options: settings for creating point cloud boundaries are now handled by an internal filter in TIndexKernel, rather than being passed as filters.hexbin options. These options are equivalent to their counterparts in filters.hexbin.

Also created new tests for the internal boundary filter in TIndexTest.cpp.

### CODE REVIEW

File: kernels/TIndexKernel.cpp [hunk 1]
```diff
@@ -654,12 +652,33 @@ pdal::Polygon TIndexKernel::prepareGeometry(const FileInfo& fileInfo)
...
     {
         SpatialReference out(m_tgtSrsString);
         g.transform(out);
     }
+
     return g;
 }
 
+
+std::string TIndexKernel::makeMultiPolygon(const std::string& wkt)
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:676]: 😬

File: kernels/TIndexKernel.hpp [hunk 2]
```diff
@@ -34,9 +34,17 @@
...
 #pragma once
 
+#include <pdal/Filter.hpp>
+#include <pdal/Streamable.hpp>
 #include <pdal/Stage.hpp>
 #include <pdal/SubcommandKernel.hpp>
 #include <pdal/util/FileUtils.hpp>
+#include <filters/private/hexer/HexGrid.hpp>
+
+namespace hexer
```
Comment (hobu) [inline kernels/TIndexKernel.hpp:44]: why do we need this forward declaration if we're also including the `HexGrid.hpp` header?

File: kernels/TIndexKernel.cpp [hunk 3]
```diff
@@ -118,6 +118,15 @@ void TIndexKernel::addSubSwitches(ProgramArgs& args,
             "Write absolute rather than relative file paths", m_absPath);
         args.add("stdin,s", "Read filespec pattern from standard input",
             m_usestdin);
+        args.add("threads", "Number of threads to use for file boundary creation",
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:121]: Maybe we should add a `requests` synonym here too.

File: kernels/TIndexKernel.cpp [hunk 4]
```diff
@@ -118,6 +118,15 @@ void TIndexKernel::addSubSwitches(ProgramArgs& args,
...
+            m_threads, 1);
+        args.add("smooth", "Smooth boundary output", m_doSmooth, true);
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:123]: Maybe this should be more accurately called `simplify`?

File: kernels/TIndexKernel.cpp [hunk 5]
```diff
@@ -118,6 +118,15 @@ void TIndexKernel::addSubSwitches(ProgramArgs& args,
...
+        args.add("threshold", "Required cell density, to be used in internal "
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:124]: `Number of points a cell must contain to be declared positive space`

File: kernels/TIndexKernel.cpp [hunk 6]
```diff
@@ -118,6 +118,15 @@ void TIndexKernel::addSubSwitches(ProgramArgs& args,
...
+            "hexbin filter (non-fast boundary)", m_density, 15);
+        args.add("edge_length", "cell edge length to be used in internal hexbin "
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:126]: This is really a `resolution`, and I think we should call it that.

File: kernels/TIndexKernel.cpp [hunk 7]
```diff
@@ -118,6 +118,15 @@ void TIndexKernel::addSubSwitches(ProgramArgs& args,
...
+            "filter (non-fast boundary)", m_edgeLength);
+        args.add("sample_size", "Sample size for auto-edge length calculation in "
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:128]: Do we have to expose this to users? They can do `--filters.hexbin.sample_size` themselves if they want.

File: kernels/TIndexKernel.cpp [hunk 8]
```diff
@@ -654,12 +652,33 @@ pdal::Polygon TIndexKernel::prepareGeometry(const FileInfo& fileInfo)
     using namespace gdal;
 
     Polygon g(fileInfo.m_boundary, fileInfo.m_srs);
+    if (fileInfo.m_gridHeight && m_doSmooth)
+    {
+        double tolerance = 1.1 * fileInfo.m_gridHeight / 2;
```
Comment (hobu) [inline kernels/TIndexKernel.cpp:657]: where did this magic come from?

**Review by hobu — CHANGES_REQUESTED** [review]

File: kernels/TIndexKernel.cpp [hunk 9]
```diff
@@ -67,6 +67,78 @@ void setDate(OGRFeatureH feature, const tm& tyme, int fieldNumber)
...
+        }
+    }
+    virtual bool processOne(PointRef& point)
+    {
+        double x = point.getFieldAs<double>(Dimension::Id::X);
+        double y = point.getFieldAs<double>(Dimension::Id::Y);
+        m_grid->addXY(x, y);
+        return true;
+    }
+    virtual void spatialReferenceChanged(const SpatialReference& srs)
```
[the remark this answers is not in the document]
Comment (ibell13) [reply kernels/TIndexKernel.cpp:125]: If I do Stage::getSpatialReference() on the filter stage without setting it explicitly, the SRS is empty. But maybe there's a different/better way to set it.

[the remark this answers is not in the document]
Comment (abellgithub) [reply kernels/TIndexKernel.cpp:125]: Yes, I forgot it was being extracted in the FileInfo.

**Review by hobu — APPROVED** [review]
